flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,463 @@
+        val secondElement = second.productElement(keyPos)
+
+        var cmp = 0
+        // both values are null -> equality
+        if (firstElement == null && secondElement == null) {
+          cmp = 0
+        }
+        // one value is null -> inequality
+        // order is considered for basic types
+        else if (firstElement == null || secondElement == null) {
+          if (comparator.isInstanceOf[BasicTypeComparator[_]]) {
+            val basicComp = comparator.asInstanceOf[BasicTypeComparator[_]]
+            if (firstElement == null) {
+              return if (basicComp.isAscendingComparison) -1 else 1
+            }
+            else if (secondElement == null) {
+              return if (basicComp.isAscendingComparison) 1 else -1
+            }
+          }
+          else {
+            return if (firstElement == null) -1 else 1
+          }
+        }
+        else {
+          cmp = comparator.compare(firstElement, secondElement)
+        }
+
+        if (cmp != 0) {
+          return cmp
+        }
+        i += 1
+      }
+    } catch {
+      case npex: NullPointerException =>
+        throw new NullKeyFieldException(keyPositions(i))
+      case iobex: IndexOutOfBoundsException =>
+        throw new KeyFieldOutOfBoundsException(keyPositions(i))
+    }
+    0
+  }
+
+  override def putNormalizedKey(record: Row, target: MemorySegment, offset: Int, numBytes: Int)
+      : Unit = {
+    var bytesLeft = numBytes
+    var currentOffset = offset
+
+    var i = 0
+    var j = 0
+    while (i < numLeadingNormalizableKeys && bytesLeft > 0) {
+      var len = normalizedKeyLengths(i)
+      len = if (bytesLeft >= len) len else bytesLeft
+
+      val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+      val element = record.productElement(keyPositions(i))
+      if (len > 0) {
+        // write a null byte with padding
+        if (element == null) {
+          target.putBoolean(currentOffset, true)
+          // write padding
+          j = 0
+          while (j < len - 1) {
+            target.put(currentOffset + j, 0.toByte)
+            j += 1
+          }
+        }
+        // write a non-null byte with key
+        else {
+          target.putBoolean(currentOffset, false)
+          // write key
+          comparator.putNormalizedKey(element, target, currentOffset, len - 1)
+        }
+      }
+
+      bytesLeft -= len
+      currentOffset += len
+      i += 1
+    }
+  }
+
+  override def extractKeys(record: Any, target: Array[AnyRef], index: Int): Int = {
+    val len = comparators.length
+    var localIndex = index
+    var i = 0
+    while (i < len) {
+      val element = record.asInstanceOf[Row].productElement(keyPositions(i))
+      localIndex += comparators(i).extractKeys(element, target, localIndex)
+      i += 1
+    }
+    localIndex - index
+  }
+
+}
+
+object RowComparator {
+  private class NullChecker extends Comparable[AnyRef] with Serializable {
```

Comment:
I don't think it is a good idea to use the `NullChecker` for reference comparisons against the field comparators. Doesn't this rely on a specific behavior of the field comparators? The field comparator might call the `equals` method on the null-valued reference object instead of the other object or it may throw a NPE if null is set as reference object.
 
I think there are two better alternatives:

1. Track which fields of the reference object are null using a Boolean array within the RowComparator and don't set null values in the field comparator.
2. Wrap each field comparator with a null-aware comparator. This comparator could also take the logic for the null handling of normalized keys.